engine/pkg/meta/internal/sqlkv/sql_impl_test.go

```diff
@@ -0,0 +1,445 @@
+	db, mock, err := sqlmock.New()
+	require.Nil(t, err)
+	// common execution for orm
+	mock.ExpectQuery("SELECT VERSION()").
+		WillReturnRows(sqlmock.NewRows([]string{"VERSION()"}).
+			AddRow("5.7.35-log"))
+	mock.ExpectQuery(regexp.QuoteMeta("SELECT SCHEMA_NAME from Information_schema.SCHEMATA " +
+		"where SCHEMA_NAME LIKE ? ORDER BY SCHEMA_NAME=? DESC limit 1")).
+		WillReturnRows(sqlmock.NewRows([]string{"SCHEMA_NAME"}))
+	mock.ExpectExec(regexp.QuoteMeta(fmt.Sprintf("CREATE TABLE `%s` (`seq_id` bigint unsigned AUTO_INCREMENT,"+
+		"`created_at` datetime(3) NULL,`updated_at` datetime(3) NULL,`key` varbinary(2048) not null,`value` blob,"+
+		"`job_id` varchar(64) not null,PRIMARY KEY (`seq_id`),UNIQUE INDEX uidx_jk (`job_id`,`key`))", table))).
+		WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectExec(regexp.QuoteMeta("INSERT INTO `logic_epoches` (`created_at`,`updated_at`,"+
+		"`job_id`,`epoch`) VALUES (?,?,?,?) ON DUPLICATE KEY UPDATE `seq_id`=`seq_id`")).
+		WithArgs(anyTime{}, anyTime{}, jobID, 1).
+		WillReturnResult(sqlmock.NewResult(1, 1))
+	return db, mock, nil
+}
+
+type anyTime struct{}
+
+func (a anyTime) Match(v driver.Value) bool {
+	_, ok := v.(time.Time)
+	return ok
+}
+
+func TestNewSQLImpl(t *testing.T) {
+	t.Parallel()
+
+	sqlDB, mock, err := mockGetDBConn(t, "test", fakeTable, fakeJob)
+
+	defer sqlDB.Close()
+	defer mock.ExpectClose()
+	require.Nil(t, err)
+	cli, err := NewSQLImpl(sqlDB, fakeTable, fakeJob)
+	defer cli.Close()
+	require.Nil(t, err)
+	require.NotNil(t, cli)
+}
+
+func TestPut(t *testing.T) {
+	t.Parallel()
+
+	sqlDB, mock, err := mockGetDBConn(t, "test", fakeTable, fakeJob)
+	defer sqlDB.Close()
+	defer mock.ExpectClose()
+	require.Nil(t, err)
+	cli, err := NewSQLImpl(sqlDB, fakeTable, fakeJob)
+	defer cli.Close()
+	require.Nil(t, err)
+	require.NotNil(t, cli)
+	var anyT anyTime
+
+	testCases := []tCase{
+		{
+			fn: "Put",
+			inputs: []interface{}{
+				"key0",
+				"value0",
+			},
+			output: &metaModel.PutResponse{
+				Header: &metaModel.ResponseHeader{},
+			},
+			mockExpectResFn: func(mock sqlmock.Sqlmock) {
+				mock.ExpectExec(regexp.QuoteMeta("INSERT INTO `fakeTable` (`created_at`,`updated_at`,"+
+					"`key`,`value`,`job_id`) VALUES (?,?,?,?,?) ON DUPLICATE KEY UPDATE `updated_at`=?,"+
+					"`key`=VALUES(`key`),`value`=VALUES(`value`),`job_id`=VALUES(`job_id`)")).
+					WithArgs(anyT, anyT, []byte("key0"), []byte("value0"), fakeJob, anyT).
+					WillReturnResult(sqlmock.NewResult(1, 1))
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		testInner(t, mock, cli, tc)
+	}
+}
+
+func TestGet(t *testing.T) {
+	t.Parallel()
+
+	sqlDB, mock, err := mockGetDBConn(t, "test", fakeTable, fakeJob)
+	defer sqlDB.Close()
+	defer mock.ExpectClose()
+	require.Nil(t, err)
+	cli, err := NewSQLImpl(sqlDB, fakeTable, fakeJob)
+	require.Nil(t, err)
+	require.NotNil(t, cli)
+
+	testCases := []tCase{
+		{
+			fn: "Get",
+			inputs: []interface{}{
+				"key0",
```

Comment:
Please add more cases of random binary bytes, not just readable ascii bytes. Maybe we can try https://go.dev/doc/fuzz/.

Reply:
Sure. But I consider to  restrict the key as acsii code.

Reply:
I works for binary bytes for byte comparation, refer to: https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
Since we can't use a mysql server in our ut, it's hard to make fuzz test here.